src/metabase/query_processor/pivot.clj

```diff
@@ -1,98 +1,144 @@
-          :when (not (zero? (bit-and (bit-shift-left 1 item) combo)))]
-      (nth xs item))))
+(defn- group-number
+  "Come up with a display name given a combination of breakout `indecies` e.g.
 
-(defn- generate-specified-breakouts
-  "Generate breakouts specified by pivot_rows"
-  [breakouts pivot-rows pivot-cols]
-  (let [generator (fn [coll]
-                    (for [pivots coll]
-                      (map #(nth breakouts %) pivots)))]
-    (concat
-     [breakouts]
-     (generator [pivot-cols])
-     (when (and (seq pivot-cols)
-                (seq pivot-rows))
-       (generator [(concat [(first pivot-rows)] pivot-cols)]))
-     (generator (powerset (or pivot-rows
-                              ;; this can happen for the public/embed endpoints,
-                              ;; where we aren't given a pivot_rows / pivot_cols
-                              ;; parameter, so we'll just generate everything
-                              (vec (range 0 (count breakouts)))))))))
+    (group-number 3 [1])   ; -> 5
+    (group-number 3 [1 2]) ; -> 1"
+  [num-breakouts indecies]
+  (transduce
+   (map (partial bit-shift-left 1))
+   (completing bit-xor)
+   (int (dec (Math/pow 2 num-breakouts)))
+   indecies))
 
-(defn add-grouping-field
+(defn- breakout-combinations
+  "Return a sequence of all breakout combinations (by index) we should generate queries for.
+
+    (breakout-combinations 3 [1 2] nil) ;; -> [[0 1 2] [] [1 2] [2] [1]]"
+  [num-breakouts pivot-rows pivot-cols]
+  ;; validate pivot-rows/pivot-cols
+  (doseq [[k pivots] {:pivot-rows pivot-rows, :pivot-cols pivot-cols}
+          i          pivots]
+    (when (>= i num-breakouts)
+      (throw (ex-info (tru "Invalid {0}: specified breakout at index {1}, but we only have {2} breakouts"
+                           (name k) i num-breakouts)
+                      {:type          qp.error-type/invalid-query
+                       :num-breakouts num-breakouts
+                       :pivot-rows    pivot-rows
+                       :pivot-cols    pivot-cols}))))
+  (sort-by
+   (partial group-number num-breakouts)
+   (distinct
+    (map
+     vec
+     (concat
+      ;; e.g. given num-breakouts = 4; pivot-rows = [0 1 2]; pivot-cols = [3]
+      ;; subtotal rows
+      ;; _.range(1, pivotRows.length).map(i => [...pivotRow.slice(0, i), ...pivotCols])
+      ;;  => [0 3] [0 1 3] => 1001 1101
+      (for [i (range 1 (count pivot-rows))]
+        (concat (take i pivot-rows) pivot-cols))
+      ;; “row totals” on the right
+      ;; pivotRows
+      ;; => [0 1 2] => 1110
+      [pivot-rows]
+      ;; subtotal rows within “row totals”
+      ;; _.range(1, pivotRows.length).map(i => pivotRow.slice(0, i))
+      ;; => [0] [0 1] => 1000 1100
+      (for [i (range 1 (count pivot-rows))]
+        (take i pivot-rows))
+      ;; “grand totals” row
+      ;; pivotCols
+      ;; => [3] => 0001
+      [pivot-cols]
+      ;; bottom right corner [] => 0000
+      [[]])))))
```

Comment:
I separated out the code that determines the combination of breakout indecies to generate queries for with the code that actually puts together the corresponding breakouts to make it easier to test and make determining the group number of the breakouts combination easier (no need to reverse the breakout back to its index later on)